### ISSUE TITLE
Функция конвертации строки

### ISSUE THREAD

**Issue by RonaldFord** [BladeMight/Mahou#252]
Попытался перейти на Mahou после использования TextCorrection от AZJIO.  И сразу бросилось в глаза, что В TextCorrection конвертация раскладки текста в выделенной области и целой строки происходит на одну и ту же клавишу, а в Mahou так сделать невозможно.

Предлагаю добавить "Функцию конвертации строки" от её начала до позиции курсора.
Без необходимости выделять текст. Без обязательного условия, что этот текст был только что набран. Просто по нажатию клавиши.
И объединить конвертацию выделенного текста и целой строки в одну функцию вызваемую единой клавишей, как сделано в TextCorrection.

**Comment by BladeMight** [comment]
> Без обязательного условия, что этот текст был только что набран. Просто по нажатию клавиши.

Так можно сделать то же что и Shift+Home => Клавиша конвертации выделенного, а если еще и:

> Без необходимости выделять текст.

Не знаю, не во всех приложениях будет работать если "не выделять"(т.е. не получать текст через буфер обмена).

Можно же просто нажимать Shift+Home и Клавиша конвертации выделенного.

P.S. Кстати для Конвертации выделенного и Конвертации последнего слова *можно* назначить на одну клавишу.

**Comment by BladeMight** [comment]
> Так можно сделать то же что и Shift+Home => Клавиша конвертации выделенного

Проверил TextCorrection, там собственно так и сделано. И в окнах где выделение не работает, и конвертация не работает, например cmd.exe.

**Comment by RonaldFord** [comment]
Может сделать функцию таким образом:
![20191211-141705_cr](https://user-images.githubusercontent.com/58736406/70618589-4e859880-1c24-11ea-8420-e4940e09a771.png)

**Comment by BladeMight** [comment]
Ну так я так и думал. Просто как и:

> P.S. Кстати для Конвертации выделенного и Конвертации последнего слова можно назначить на одну клавишу.

Просто с условием что был введен текст или нет(выделить всю строку, если уже не выделено что-то).

**Comment by RonaldFord** [comment]
Здравствуйте, BladeMight. Вы так и не добавили функцию, которая мне очень нужна. Хотя позиционируете проект как самый гибкий - "Mahou работает так как Вы хотите, настраивайте как пожелаете". И не сообщили о своём решении, реализовать или отклонить запрос. Вы посчитали функцию излишней или не поняли её суть? Я могу дать пояснения.

**Comment by BladeMight** [comment]
@RonaldFord Привет, пока что можно сделать что-то похожее с использованием [Mahou.mm](https://github.com/BladeMight/Mahou/wiki/Mahou.mm_RU), см. https://github.com/BladeMight/Mahou/issues/296#issuecomment-717597197

- Я так понял ты хочешь чтобы 1 горячая клавиша работала как 3? Т.е. конвертация *последнего* слова если ты вводил, если нет значит выделенного, а если нет *выделенного* тогда выделить с помощью `Shift+Home` и *конвертировать линию*. В итоге 3 горячих клавиши одной так?

**Comment by RonaldFord** [comment]
Да, верно. Только конвертация всей введённой строки (несколько слов). Так результат будет ближе к функции "Shift+Home и конвертировать линию".  По моему ощущению удобнее вводить фразу до конца, и уже потом её исправлять, чем вовремя останавливаться.

Кстати, где-то есть настройка задержки между возвращаемыми клавишами? Хотелось бы настроить вывод близкий к моментальному.

**Comment by BladeMight** [comment]
>Кстати, где-то есть настройка задержки между возвращаемыми клавишами?

Для функции `kbd` как и `__keyboard` в cниппетах, если указать: `|число` - будет использоваться задержка между каждым посыланием нажатия клавиш, однако в Mahou.mm `|` используется для разделения, нужно будет придумать другой способ...

**Comment by BladeMight** [comment]
@RonaldFord Добавил в 5851863 новое действие для `Mahou.mm`, которое работает только в `multi`, это действие `ifhk` - проверяет успешно ли выполнена горячая клавиша конвертации, работает вот так: `ifhk|n` - значит если последняя клавиша не успешно сработала тогда продолжать `multi`, или вот `ifhk|y` - если последняя клавиша конвертации успешно сработала, твоя тройная горячая клавиша будет такой:

```
TripleConvert|^^f9|multi|hk|c0|ifhk|n|hk|s7|ifhk|n|kbd|lshift+home|hk|s7
```

объясняю:

```
multi:
  - hk|c0 = конвертация последнего слова
  - ifhk|n = если успешна, не продолжает, иначе:
  - - hk|s7 = конвертация выделения
  - - ifhk|n = если успешна, не продолжает, иначе:
  - - - kbd|lshift+home = выделяем линию с помощью lshift+home
  - - - hk|s7 = конвертация выделения
```

Еще нужно wiki обновить...

**Comment by RonaldFord** [comment]
Уже можно проверить?

**Comment by BladeMight** [comment]
> Это нужно добавить в сниппеты?

Нет в `Mahou.mm`, см. [wiki=>Mahou.mm](https://github.com/BladeMight/Mahou/wiki/Mahou.mm_RU)

> Для функции kbd как и __keyboard в cниппетах, если указать: |число - будет использоваться задержка между каждым посыланием нажатия клавиш, однако в Mahou.mm | используется для разделения, нужно будет придумать другой способ...

Добавил в 70575de.

**Comment by BladeMight** [comment]
Проверяй в latest-commit.

Для обновления до latest-commit:
1. Во вкладке "Обновления" поставить канал обновлений на LatestCommit(Beta), и обновиться как обычно(проверить, скачать обновление)
2. 2. На странице: https://github.com/BladeMight/Mahou/releases/tag/latest-commit есть постоянно авто-обновляемые сборки. (если не знаешь какую брать - Release_x86_x64.zip)

**Comment by RonaldFord** [comment]
Программу обновил. Разбираюсь с синтаксисом Mahou.mm

**Comment by RonaldFord** [comment]
Работает отлично. Всё как задумывалось.
В файл написал такой текст:
TripleConvert|^^f9|multi|hk|c1|ifhk|n|hk|s7|ifhk|n|kbd|lshift+home|hk|s7

Куда поставить задержки мс.? Написал в конец !!500. Должна быть задержка пол секунды между каждым символом. Но функция работает как и раньше

**Comment by BladeMight** [comment]
> Куда поставить задержки мс.? Написал в конец !!500

Задержки только для функции `kbd`, а в этой *конструкции* только для `lshift+home`, и ставить нужно в конец аргумента для функции `kbd`, т.е. после `lshift+home` => `lshift+home!!500`.

**Comment by BladeMight** [comment]
> Должна быть задержка пол секунды между каждым символом

В каком месте ты хочешь задержку?

**Comment by BladeMight** [comment]
@RonaldFord 

> Еще нужно wiki обновить...

Обновил, [wiki=>Mahou.mm](https://github.com/BladeMight/Mahou/wiki/Mahou.mm_RU)

**Comment by RonaldFord** [comment]
Задержка 500мс - это был эксперимент, чтобы найти аргумент управляющий задержкой.
Чтобы затем поставить её на минимум. Задержка между каждым эмулированным нажатием клавиш по умолчанию минимальна, или есть некоторое значение?

Попробовал так:
TripleConvert|^^Pause|multi|hk|c1|ifhk|n|hk|s7|ifhk|n|kbd|lshift+home!!500|hk|s7
Конструкция даёт две задержки. После нажатия клавиши и после выделения текста. На скорость вывода символов это не повлияло.
Это не тот результат, которого я ожидал.

**Comment by BladeMight** [comment]
> Конструкция даёт две задержки

Ну так там две клавиши `lshift` и `home`, сначала нажимает `lshift`, ждёт 500мс, затем нажимает `home` и опять ждёт 500мс.

>  Задержка между каждым эмулированным нажатием клавиш по умолчанию минимальна, или есть некоторое значение?

По умолчанию: **0**мс. ограничивается самой скоростью функции WinAPI: "SendInput". (см. [ss](https://www.google.com/search?q=SendInput+slow))

> На скорость вывода символов

Ты имеешь в виду вывод символов от функции "конвертация выделения"(hk|s7)? Если да, то это скорость работы `SendInput` и с этим ничего не сделаешь, однако в Mahou еще можно использовать вместо "ввода" - вставку текста:
![2020 12 28 - 21'24'50_rect](https://user-images.githubusercontent.com/11931888/103238459-ddfb0800-495b-11eb-9f1b-216522797f46.jpg)
